src/app.ts

```diff
@@ -95,79 +97,89 @@ if (cluster.isMaster) {
+	masterLogger.debug('registering as service instance...');
 	service.wrap(serviceInstance => {
-		logger.info(
+		const serviceLogger = getLogger('master', serviceInstance.getId());
+		serviceLogger.info(
 			`registered as service instance with id=${serviceInstance.getId()}`,
 		);
-		logger.info(
-			`spawning ${VPN_INSTANCE_COUNT} worker${
-				VPN_INSTANCE_COUNT > 1 ? 's' : ''
-			}`,
-		);
-		_.times(VPN_INSTANCE_COUNT, i => {
-			const workerId = i + 1;
-			const restartWorker = (code?: number, signal?: string) => {
-				if (signal != null) {
-					logger.crit(`worker-${workerId} killed with signal ${signal}`);
-				}
-				if (code != null) {
-					logger.crit(`worker-${workerId} exited with code ${code}`);
-				}
-				const env = {
-					...process.env,
-					WORKER_ID: workerId,
-					SERVICE_ID: serviceInstance.getId(),
-					VPN_VERBOSE_LOGS: verbose,
+
+		serviceLogger.info('spawning vpn authentication api server...');
+		const api = apiServer(serviceInstance.getId());
+		api.listenAsync(VPN_API_PORT).then(() => {
+			serviceLogger.info(
+				`spawning ${VPN_INSTANCE_COUNT} worker${
+					VPN_INSTANCE_COUNT > 1 ? 's' : ''
+				}`,
+			);
+			_.times(VPN_INSTANCE_COUNT, i => {
+				const workerId = i + 1;
+				const restartWorker = (code?: number, signal?: string) => {
+					if (signal != null) {
+						serviceLogger.crit(
+							`worker-${workerId} killed with signal ${signal}`,
+						);
+					}
+					if (code != null) {
+						serviceLogger.crit(`worker-${workerId} exited with code ${code}`);
+					}
+					const env = {
+						...process.env,
+						WORKER_ID: workerId,
+						SERVICE_ID: serviceInstance.getId(),
+						VPN_VERBOSE_LOGS: verbose,
+					};
+					cluster.fork(env).on('exit', restartWorker);
 				};
-				cluster.fork(env).on('exit', restartWorker);
-			};
-			restartWorker();
+				restartWorker();
+			});
+
+			const app = express();
+			app.disable('x-powered-by');
+			app.get('/ping', (_req, res) => res.send('OK'));
+			app
+				.get('/cluster_metrics', (_req, res) => {
+					for (const clientMetrics of Object.values(workerMetrics)) {
+						metrics.histogram(
+							Metrics.SessionRxBitrate,
+							_.mean(clientMetrics.rxBitrate),
+						);
+						metrics.histogram(
+							Metrics.SessionTxBitrate,
+							_.mean(clientMetrics.txBitrate),
+						);
+					}
+					return new prometheus.AggregatorRegistry()
+						.clusterMetrics()
+						.then((clusterMetrics: string) => {
+							res.set('Content-Type', prometheus.register.contentType);
+							res.write(prometheus.register.metrics());
+							res.write('\n');
+							res.write(clusterMetrics);
+							res.end();
+							workerMetrics = {};
+							metrics.reset(Metrics.SessionRxBitrate);
+							metrics.reset(Metrics.SessionTxBitrate);
+						})
+						.catch((err: Error) => {
+							serviceLogger.warning(`error in /cluster_metrics: ${err}`);
+							res.status(500).send();
+						});
+				})
+				.listen(8080);
+
+			return Promise.all([app, metrics]);
```

Comment:
Aren't both of those synchronous values?  In which case this would be better as:
```suggestion
			return [app, metrics];
```

Reply:
Good point, I was thinking about the `listenAsync` return value.